<a href="https://colab.research.google.com/github/Shrason/Projects/blob/main/Data_cleaning_TSV_File.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('/content/chipotle.tsv',sep='\t')

In [3]:
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [4]:
df.tail()

order_id  quantity           item_name  \
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

Missing values and data types

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


Notes:
- choice_description column has 1246 missing values

In [6]:
df.describe()

order_id     quantity
count  4622.000000  4622.000000
mean    927.254868     1.075725
std     528.890796     0.410186
min       1.000000     1.000000
25%     477.250000     1.000000
50%     926.000000     1.000000
75%    1393.000000     1.000000
max    1834.000000    15.000000

Duplicate data

In [7]:
df['item_name'].duplicated().any()

True

In [8]:
df['order_id'].duplicated().any()

True

- As order id is always incremental.
- In the oder_id column the max value is 1834 , where as the data size is 4622.
-  Thus, duplicacy is removed

In [9]:
df = df.drop_duplicates(subset=['order_id'])

In [10]:
df['order_id'].duplicated().any()

False

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1834 entries, 0 to 4619
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            1834 non-null   int64 
 1   quantity            1834 non-null   int64 
 2   item_name           1834 non-null   object
 3   choice_description  1696 non-null   object
 4   item_price          1834 non-null   object
dtypes: int64(2), object(3)
memory usage: 86.0+ KB


Quantity and Item Price

In [12]:
df['item_price'] = df['item_price'].str.replace('$', '')
df['item_price'].head()

<ipython-input-12-4bde2f3b9c26>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['item_price'] = df['item_price'].str.replace('$', '')


0     2.39 
4    16.98 
5    10.98 
7    11.75 
9     9.25 
Name: item_price, dtype: object

In [13]:
df['item_price'] = df['item_price'].astype(float)

Choice Description

In [14]:
df['choice_description'].head()

0                                                  NaN
4    [Tomatillo-Red Chili Salsa (Hot), [Black Beans...
5    [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...
7    [Tomatillo Red Chili Salsa, [Fajita Vegetables...
9    [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...
Name: choice_description, dtype: object

In [15]:
df['choice_description'].tail()

4610    [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...
4612    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...
4615     [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]
4617    [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...
4619    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
Name: choice_description, dtype: object

- In the choice_description column we have list and some nested list.
- The list is of the ingredients specified for the dish, can be joined to one list.
- The column also conatains NaN value which can be replace with 'No Choice'.


In [16]:
#df['choice_description'] = df['choice_description'].str.replace(r'\[\]','')
#df['choice_description'] = df['choice_description'].fillna('No choice')
#df['choice_description'].head()

Item Name Standardization

In [17]:
df['item_name'].head()

0    Chips and Fresh Tomato Salsa
4                    Chicken Bowl
5                    Chicken Bowl
7                   Steak Burrito
9                   Steak Burrito
Name: item_name, dtype: object

In [18]:
df['item_name'].tail()

4610         Steak Burrito
4612         Carnitas Bowl
4615    Chicken Soft Tacos
4617         Steak Burrito
4619    Chicken Salad Bowl
Name: item_name, dtype: object

Quantity and Price Relationships

In [19]:
correlation = df['quantity'].corr(df['item_price'])
print(correlation)

0.4868381407447502


- This shows a positve relation between Quantity and Price Relationships

Converting to CSV

In [20]:

csv_file = 'chipotle.csv'
df.to_csv(csv_file)

In [21]:
df.head()

order_id  quantity                     item_name  \
0         1         1  Chips and Fresh Tomato Salsa   
4         2         2                  Chicken Bowl   
5         3         1                  Chicken Bowl   
7         4         1                 Steak Burrito   
9         5         1                 Steak Burrito   

                                  choice_description  item_price  
0                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        9.25

Quantity of each item


In [25]:
item = df.groupby('item_name')
item_quantity = item['quantity'].sum()
item_quantity

item_name
6 Pack Soft Drink                         11
Barbacoa Bowl                             41
Barbacoa Burrito                          51
Barbacoa Crispy Tacos                      3
Barbacoa Salad Bowl                        3
Barbacoa Soft Tacos                       11
Bottled Water                             33
Bowl                                       4
Burrito                                    3
Canned Soda                               15
Canned Soft Drink                         48
Carnitas Bowl                             38
Carnitas Burrito                          36
Carnitas Crispy Tacos                      2
Carnitas Salad                             1
Carnitas Salad Bowl                        1
Carnitas Soft Tacos                       13
Chicken Bowl                             501
Chicken Burrito                          369
Chicken Crispy Tacos                      12
Chicken Salad                              6
Chicken Salad Bowl                        72


What was the most ordered item in the choice_description column?

In [26]:
df2 = df['choice_description'].str.split(',', expand = True).stack().reset_index(drop=True).to_frame('choice_description')
df2['choice_description'] = df2['choice_description'].str.strip()
most_ordered = df2['choice_description'].value_counts()
most_ordered

Cheese                                  1076
Lettuce]]                                850
[Rice                                    819
Sour Cream                               737
[Fresh Tomato Salsa                      591
                                        ... 
[Tomatillo Green Chili Salsa]              1
[Roasted Chili Corn Salsa]                 1
[Braised Barbacoa                          1
[Adobo-Marinated and Grilled Chicken       1
[Brown Rice]                               1
Name: choice_description, Length: 91, dtype: int64

How many items were ordered in total?


In [30]:
total_order = item_quantity.sum()
total_order

1993

How much was the revenue for the period in the dataset?


In [42]:
df['revenue'] = df['item_price']*df['quantity']
df['revenue'].sum()

20732.910000000003

How many orders were made in the period?


In [35]:
df['order_id'].max()

1834

What is the average revenue amount per order?


In [43]:
grouped = df.groupby("order_id")["revenue"].sum()
grouped.mean()

11.304749182115597

How many different items are sold?

In [49]:
df['item_name'].nunique()

45